File: android/app/src/main/java/app/nearboard/app/MeshPeripheralPlugin.java
package app.nearboard.app;

import android.Manifest;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattServer;
import android.bluetooth.BluetoothGattServerCallback;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothManager;
import android.bluetooth.le.AdvertiseCallback;
import android.bluetooth.le.AdvertiseData;
import android.bluetooth.le.AdvertiseSettings;
import android.bluetooth.le.BluetoothLeAdvertiser;
import android.content.Context;
import android.os.Build;
import android.os.ParcelUuid;
import android.util.Base64;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * The peripheral half of the mesh.
 *
 * @capacitor-community/bluetooth-le is central-only — it can connect out but
 * cannot advertise or serve GATT. A mesh needs both roles on every device, so
 * this plugin supplies advertising and the GATT server; the community plugin
 * still handles the central side.
 *
 * Android specifics worth knowing:
 *
 *  - BLUETOOTH_ADVERTISE is required from API 31 (declared in the manifest).
 *  - Sustained scanning and advertising in the background needs a foreground
 *    service, which this plugin does not start. Backgrounded, the mesh will go
 *    quiet — a deliberate scope limit, not an oversight.
 *  - Many chipsets cap concurrent GATT connections around 4-7.
 *  - The advertisement payload is 31 bytes. A 128-bit service UUID uses 18 of
 *    them, so the device name is sent in the scan response rather than the
 *    advertisement, which would otherwise fail to register.
 */
@CapacitorPlugin(
    name = "MeshPeripheral",
    permissions = {
        @Permission(alias = "advertise", strings = { Manifest.permission.BLUETOOTH_ADVERTISE }),
        @Permission(alias = "connect", strings = { Manifest.permission.BLUETOOTH_CONNECT })
    }
)
public class MeshPeripheralPlugin extends Plugin {

    private static final UUID CCC_DESCRIPTOR_UUID =
        UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");

    private BluetoothManager bluetoothManager;
    private BluetoothLeAdvertiser advertiser;
    private BluetoothGattServer gattServer;

    private BluetoothGattCharacteristic outboundCharacteristic;

    /** Centrals subscribed to notifications, keyed by address. */
    private final Map<String, BluetoothDevice> subscribers = new HashMap<>();
    /** Negotiated usable payload per device. */
    private final Map<String, Integer> deviceMtu = new HashMap<>();

    private AdvertiseCallback advertiseCallback;
    private PluginCall startCall;

    @Override
    public void load() {
        bluetoothManager = (BluetoothManager) getContext().getSystemService(Context.BLUETOOTH_SERVICE);
    }

    // ---- Plugin API ----

    @PluginMethod
    public void startAdvertising(PluginCall call) {
        String service = call.getString("serviceUuid");
        String inbound = call.getString("inboundUuid");
        String outbound = call.getString("outboundUuid");

        if (service == null || inbound == null || outbound == null) {
            call.reject("serviceUuid, inboundUuid and outboundUuid are required");
            return;
        }

        if (bluetoothManager == null || bluetoothManager.getAdapter() == null) {
            call.reject("Bluetooth is not available on this device");
            return;
        }
        if (!bluetoothManager.getAdapter().isEnabled()) {
            call.reject("Bluetooth is turned off");
            return;
        }

        advertiser = bluetoothManager.getAdapter().getBluetoothLeAdvertiser();
        if (advertiser == null) {
            // Not every chipset supports the peripheral role, and there is no
            // way to know before asking.
            call.reject("This device cannot advertise as a BLE peripheral");
            return;
        }

        try {
            startGattServer(UUID.fromString(service), UUID.fromString(inbound), UUID.fromString(outbound));
        } catch (IllegalArgumentException e) {
            call.reject("Invalid UUID: " + e.getMessage());
            return;
        }

        startCall = call;
        call.setKeepAlive(true);
        beginAdvertising(UUID.fromString(service));
    }

    @PluginMethod
    public void stopAdvertising(PluginCall call) {
        if (advertiser != null && advertiseCallback != null) {
            advertiser.stopAdvertising(advertiseCallback);
            advertiseCallback = null;
        }
        if (gattServer != null) {
            gattServer.close();
            gattServer = null;
        }
        subscribers.clear();
        deviceMtu.clear();
        call.resolve();
    }

    /**
     * Push one frame to subscribed centrals. `centralId` targets a single peer;
     * omitting it broadcasts. The mesh targets, so a relay does not echo a
     * packet back to the peer it came from.
     */
    @PluginMethod
    public void notify(PluginCall call) {
        String base64 = call.getString("data");
        if (base64 == null) {
            call.reject("data is required");
            return;
        }
        if (gattServer == null || outboundCharacteristic == null) {
            call.reject("Not advertising");
            return;
        }

        byte[] payload;
        try {
            payload = Base64.decode(base64, Base64.NO_WRAP);
        } catch (IllegalArgumentException e) {
            call.reject("data must be valid base64");
            return;
        }

        List<BluetoothDevice> targets = new ArrayList<>();
        String centralId = call.getString("centralId");
        if (centralId != null) {
            BluetoothDevice device = subscribers.get(centralId);
            if (device == null) {
                call.reject("Unknown central " + centralId);
                return;
            }
            targets.add(device);
        } else {
            targets.addAll(subscribers.values());
        }

        int sent = 0;
        for (BluetoothDevice device : targets) {
            if (sendNotification(device, payload)) sent++;
        }

        JSObject result = new JSObject();
        result.put("sent", sent);
        call.resolve(result);
    }

    @PluginMethod
    public void getSubscribers(PluginCall call) {
        JSObject result = new JSObject();
        result.put("centralIds", new ArrayList<>(subscribers.keySet()));
        call.resolve(result);
    }

    // ---- GATT server ----

    private void startGattServer(UUID serviceUuid, UUID inboundUuid, UUID outboundUuid) {
        gattServer = bluetoothManager.openGattServer(getContext(), gattServerCallback);

        // Inbound: peers write frames to us. WRITE_NO_RESPONSE avoids an ack per
        // fragment, which would dominate transfer time for a large post.
        BluetoothGattCharacteristic inbound = new BluetoothGattCharacteristic(
            inboundUuid,
            BluetoothGattCharacteristic.PROPERTY_WRITE
                | BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE,
            BluetoothGattCharacteristic.PERMISSION_WRITE
        );

        // Outbound: we push frames to peers. Notify inverts the direction a GATT
        // read allows, so a peripheral can initiate.
        outboundCharacteristic = new BluetoothGattCharacteristic(
            outboundUuid,
            BluetoothGattCharacteristic.PROPERTY_NOTIFY,
            BluetoothGattCharacteristic.PERMISSION_READ
        );

        // Android requires the Client Characteristic Configuration descriptor to
        // exist before a central can subscribe. Omitting it is a common cause of
        // notifications that register successfully and then never arrive.
        BluetoothGattDescriptor ccc = new BluetoothGattDescriptor(
            CCC_DESCRIPTOR_UUID,
            BluetoothGattDescriptor.PERMISSION_READ | BluetoothGattDescriptor.PERMISSION_WRITE
        );
        outboundCharacteristic.addDescriptor(ccc);

        BluetoothGattService service =
            new BluetoothGattService(serviceUuid, BluetoothGattService.SERVICE_TYPE_PRIMARY);
        service.addCharacteristic(inbound);
        service.addCharacteristic(outboundCharacteristic);

        gattServer.addService(service);
    }

    private void beginAdvertising(UUID serviceUuid) {
        AdvertiseSettings settings = new AdvertiseSettings.Builder()
            .setAdvertiseMode(AdvertiseSettings.ADVERTISE_MODE_BALANCED)
            .setTxPowerLevel(AdvertiseSettings.ADVERTISE_TX_POWER_MEDIUM)
            .setConnectable(true)
            .setTimeout(0) // advertise until told to stop
            .build();

        // 31-byte budget: a 128-bit UUID takes 18, so the name goes in the scan
        // response. Including both here makes registration fail outright.
        AdvertiseData data = new AdvertiseData.Builder()
            .setIncludeDeviceName(false)
            .addServiceUuid(new ParcelUuid(serviceUuid))
            .build();

        AdvertiseData scanResponse = new AdvertiseData.Builder()
            .setIncludeDeviceName(true)
            .build();

        advertiseCallback = new AdvertiseCallback() {
            @Override
            public void onStartSuccess(AdvertiseSettings settingsInEffect) {
                resolveStart();
            }

            @Override
            public void onStartFailure(int errorCode) {
                rejectStart("Advertising failed: " + describeAdvertiseError(errorCode));
            }
        };

        advertiser.startAdvertising(settings, data, scanResponse, advertiseCallback);
    }

    private final BluetoothGattServerCallback gattServerCallback = new BluetoothGattServerCallback() {

        @Override
        public void onConnectionStateChange(BluetoothDevice device, int status, int newState) {
            String address = device.getAddress();
            if (newState == BluetoothGatt.STATE_DISCONNECTED) {
                subscribers.remove(address);
                deviceMtu.remove(address);
                JSObject event = new JSObject();
                event.put("centralId", address);
                notifyListeners("peerDisconnected", event);
            }
        }

        @Override
        public void onMtuChanged(BluetoothDevice device, int mtu) {
            // Reported MTU includes the 3-byte ATT header, which is not usable
            // payload. Store the usable figure so JS fragments to the right size.
            deviceMtu.put(device.getAddress(), Math.max(1, mtu - 3));
        }

        @Override
        public void onCharacteristicWriteRequest(
            BluetoothDevice device,
            int requestId,
            BluetoothGattCharacteristic characteristic,
            boolean preparedWrite,
            boolean responseNeeded,
            int offset,
            byte[] value
        ) {
            if (value != null && value.length > 0) {
                JSObject event = new JSObject();
                event.put("centralId", device.getAddress());
                event.put("data", Base64.encodeToString(value, Base64.NO_WRAP));
                notifyListeners("frameReceived", event);
            }

            if (responseNeeded && gattServer != null) {
                gattServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, null);
            }
        }

        @Override
        public void onDescriptorWriteRequest(
            BluetoothDevice device,
            int requestId,
            BluetoothGattDescriptor descriptor,
            boolean preparedWrite,
            boolean responseNeeded,
            int offset,
            byte[] value
        ) {
            // A central subscribing writes ENABLE_NOTIFICATION_VALUE to the CCC
            // descriptor. This is the only signal that it is ready to receive.
            if (CCC_DESCRIPTOR_UUID.equals(descriptor.getUuid())) {
                boolean enabling = value != null
                    && value.length > 0
                    && value[0] == BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE[0];

                String address = device.getAddress();
                if (enabling) {
                    subscribers.put(address, device);
                    JSObject event = new JSObject();
                    event.put("centralId", address);
                    event.put("mtu", deviceMtu.containsKey(address) ? deviceMtu.get(address) : 20);
                    notifyListeners("peerConnected", event);
                } else {
                    subscribers.remove(address);
                    JSObject event = new JSObject();
                    event.put("centralId", address);
                    notifyListeners("peerDisconnected", event);
                }
            }

            if (responseNeeded && gattServer != null) {
                gattServer.sendResponse(device, requestId, BluetoothGatt.GATT_SUCCESS, offset, null);
            }
        }
    };

    @SuppressWarnings("deprecation")
    private boolean sendNotification(BluetoothDevice device, byte[] payload) {
        if (gattServer == null || outboundCharacteristic == null) return false;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            // API 33+ takes the value directly, avoiding the shared-state race
            // in the older setValue-then-notify pattern.
            return gattServer.notifyCharacteristicChanged(device, outboundCharacteristic, false, payload)
                == BluetoothGatt.GATT_SUCCESS;
        }

        outboundCharacteristic.setValue(payload);
        return gattServer.notifyCharacteristicChanged(device, outboundCharacteristic, false);
    }

    // ---- Helpers ----

    private void resolveStart() {
        if (startCall == null) return;
        startCall.resolve();
        startCall.setKeepAlive(false);
        startCall = null;
    }

    private void rejectStart(String message) {
        if (startCall == null) return;
        startCall.reject(message);
        startCall.setKeepAlive(false);
        startCall = null;
    }

    private static String describeAdvertiseError(int code) {
        switch (code) {
            case AdvertiseCallback.ADVERTISE_FAILED_DATA_TOO_LARGE:
                return "advertisement payload exceeds 31 bytes";
            case AdvertiseCallback.ADVERTISE_FAILED_TOO_MANY_ADVERTISERS:
                return "too many advertisers already running";
            case AdvertiseCallback.ADVERTISE_FAILED_ALREADY_STARTED:
                return "already advertising";
            case AdvertiseCallback.ADVERTISE_FAILED_INTERNAL_ERROR:
                return "internal Bluetooth stack error";
            case AdvertiseCallback.ADVERTISE_FAILED_FEATURE_UNSUPPORTED:
                return "peripheral role unsupported on this chipset";
            default:
                return "unknown error " + code;
        }
    }
}

File: ios/App/App/MeshPeripheralPlugin.swift
import Foundation
import Capacitor
import CoreBluetooth

/**
 * The peripheral half of the mesh.
 *
 * @capacitor-community/bluetooth-le is central-only: it can connect out to
 * something that advertises, but it cannot advertise or serve GATT. That is
 * fine for talking to a dedicated hub and useless for a mesh, where every
 * device must be discoverable by every other. This plugin supplies the missing
 * half; the community plugin still handles the central role.
 *
 * Known iOS constraints, none of which have workarounds:
 *
 *  - Backgrounded, iOS moves the service UUID into the advertisement "overflow
 *    area". Only another iOS device explicitly scanning for that exact UUID can
 *    see it; Android cannot see it at all. Background discovery is therefore
 *    degraded by design, and the app is strongest in the foreground.
 *  - The local name is not advertised in the background.
 *  - Requires `bluetooth-peripheral` in UIBackgroundModes (already set).
 */
@objc(MeshPeripheralPlugin)
public class MeshPeripheralPlugin: CAPPlugin, CBPeripheralManagerDelegate {

    private var manager: CBPeripheralManager?
    private var inboundCharacteristic: CBMutableCharacteristic?
    private var outboundCharacteristic: CBMutableCharacteristic?

    /// Centrals currently subscribed to notifications, keyed by their identifier.
    private var subscribers: [String: CBCentral] = [:]

    /// Notifications iOS refused to send because its queue was full, retried on
    /// `peripheralManagerIsReady`. Without this, a large post silently loses
    /// fragments the moment the queue backs up — which it will, since a 150 KB
    /// post is over a thousand frames.
    private var pendingNotifications: [(data: Data, central: CBCentral?)] = []

    private var serviceUUID: CBUUID?
    private var inboundUUID: CBUUID?
    private var outboundUUID: CBUUID?
    private var localName: String = "nearboard"

    private var startCall: CAPPluginCall?
    private var serviceAdded = false

    // MARK: - Plugin API

    /// Begin advertising and stand up the GATT server.
    @objc func startAdvertising(_ call: CAPPluginCall) {
        guard let service = call.getString("serviceUuid"),
              let inbound = call.getString("inboundUuid"),
              let outbound = call.getString("outboundUuid") else {
            call.reject("serviceUuid, inboundUuid and outboundUuid are required")
            return
        }

        serviceUUID = CBUUID(string: service)
        inboundUUID = CBUUID(string: inbound)
        outboundUUID = CBUUID(string: outbound)
        localName = call.getString("localName") ?? "nearboard"

        // Held until the radio reports poweredOn — resolving now would tell the
        // caller we are advertising before we are.
        startCall = call
        call.keepAlive(true)

        if manager == nil {
            manager = CBPeripheralManager(delegate: self, queue: nil, options: [
                CBPeripheralManagerOptionShowPowerAlertKey: true
            ])
        } else if manager?.state == .poweredOn {
            configureAndAdvertise()
        }
    }

    @objc func stopAdvertising(_ call: CAPPluginCall) {
        manager?.stopAdvertising()
        manager?.removeAllServices()
        serviceAdded = false
        subscribers.removeAll()
        pendingNotifications.removeAll()
        call.resolve()
    }

    /**
     * Push one frame to subscribed centrals.
     *
     * `centralId` targets a single peer; omitting it broadcasts. The mesh uses
     * the targeted form so a relay does not echo a packet back to the peer it
     * came from.
     */
    @objc func notify(_ call: CAPPluginCall) {
        guard let base64 = call.getString("data"),
              let data = Data(base64Encoded: base64) else {
            call.reject("data must be a base64 string")
            return
        }
        guard let characteristic = outboundCharacteristic else {
            call.reject("Not advertising")
            return
        }

        var targets: [CBCentral]? = nil
        if let centralId = call.getString("centralId") {
            guard let central = subscribers[centralId] else {
                call.reject("Unknown central \(centralId)")
                return
            }
            targets = [central]
        }

        let sent = manager?.updateValue(data, for: characteristic, onSubscribedCentrals: targets) ?? false
        if !sent {
            // Queue full. iOS calls peripheralManagerIsReady when it drains.
            pendingNotifications.append((data: data, central: targets?.first))
        }

        call.resolve(["queued": !sent])
    }

    @objc func getSubscribers(_ call: CAPPluginCall) {
        call.resolve(["centralIds": Array(subscribers.keys)])
    }

    // MARK: - CBPeripheralManagerDelegate

    public func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {
        switch peripheral.state {
        case .poweredOn:
            configureAndAdvertise()
        case .unauthorized:
            rejectStart("Bluetooth permission denied")
        case .unsupported:
            rejectStart("BLE peripheral role not supported on this device")
        case .poweredOff:
            notifyListeners("stateChange", data: ["state": "poweredOff"])
        default:
            break
        }
    }

    private func configureAndAdvertise() {
        guard let manager = manager,
              let serviceUUID = serviceUUID,
              let inboundUUID = inboundUUID,
              let outboundUUID = outboundUUID else { return }

        if !serviceAdded {
            // Inbound: peers write frames to us. writeWithoutResponse avoids an
            // ack per fragment, which would dominate transfer time.
            let inbound = CBMutableCharacteristic(
                type: inboundUUID,
                properties: [.write, .writeWithoutResponse],
                value: nil,
                permissions: [.writeable]
            )

            // Outbound: we push frames to peers. Notify inverts the direction a
            // GATT read allows, so a peripheral with something to say can say it.
            let outbound = CBMutableCharacteristic(
                type: outboundUUID,
                properties: [.notify],
                value: nil,
                permissions: [.readable]
            )

            let service = CBMutableService(type: serviceUUID, primary: true)
            service.characteristics = [inbound, outbound]

            inboundCharacteristic = inbound
            outboundCharacteristic = outbound

            manager.add(service)
            serviceAdded = true
        }

        manager.startAdvertising([
            CBAdvertisementDataServiceUUIDsKey: [serviceUUID],
            CBAdvertisementDataLocalNameKey: localName
        ])
    }

    public func peripheralManagerDidStartAdvertising(_ peripheral: CBPeripheralManager, error: Error?) {
        if let error = error {
            rejectStart("Failed to advertise: \(error.localizedDescription)")
            return
        }
        resolveStart()
    }

    public func peripheralManager(_ peripheral: CBPeripheralManager,
                                  central: CBCentral,
                                  didSubscribeTo characteristic: CBCharacteristic) {
        let id = central.identifier.uuidString
        subscribers[id] = central
        notifyListeners("peerConnected", data: [
            "centralId": id,
            // Usable notification payload, i.e. MTU minus the ATT header. This
            // is the fragment budget the JS layer must respect.
            "mtu": central.maximumUpdateValueLength
        ])
    }

    public func peripheralManager(_ peripheral: CBPeripheralManager,
                                  central: CBCentral,
                                  didUnsubscribeFrom characteristic: CBCharacteristic) {
        let id = central.identifier.uuidString
        subscribers.removeValue(forKey: id)
        notifyListeners("peerDisconnected", data: ["centralId": id])
    }

    public func peripheralManager(_ peripheral: CBPeripheralManager,
                                  didReceiveWrite requests: [CBATTRequest]) {
        for request in requests {
            guard let value = request.value else { continue }
            notifyListeners("frameReceived", data: [
                "centralId": request.central.identifier.uuidString,
                "data": value.base64EncodedString()
            ])
        }
        // Only the first request needs a response, and only for .write.
        if let first = requests.first {
            peripheral.respond(to: first, withResult: .success)
        }
    }

    /// iOS drained its notification queue; retry what it refused earlier.
    public func peripheralManagerIsReady(toUpdateSubscribers peripheral: CBPeripheralManager) {
        guard let characteristic = outboundCharacteristic else { return }

        while !pendingNotifications.isEmpty {
            let next = pendingNotifications[0]
            let targets = next.central.map { [$0] }
            let sent = peripheral.updateValue(next.data, for: characteristic, onSubscribedCentrals: targets)
            if !sent { return } // still full; wait for the next callback
            pendingNotifications.removeFirst()
        }
    }

    // MARK: - Helpers

    private func resolveStart() {
        startCall?.resolve()
        startCall?.keepAlive(false)
        startCall = nil
    }

    private func rejectStart(_ message: String) {
        startCall?.reject(message)
        startCall?.keepAlive(false)
        startCall = nil
        notifyListeners("stateChange", data: ["state": "error", "message": message])
    }
}
